<a href="https://colab.research.google.com/github/ArifAygun/Magnimind-Jan2023-Assignments/blob/main/Pandas_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas Assignment - Arif Aygun

In [1]:
import pandas as pd

### 1- How many rows of data are in the DataFrame?

Please refer to the following link for this practice:

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv

In [ ]:
ks_project = pd.read_csv(r"https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv")
ks_project.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

In [ ]:
ks_project.shape

(65418, 13)

DataFrame has 65418 rows

### 2- What are the names and data types of the columns?

In [ ]:
print(ks_project.columns)

Index(['id', 'name', 'category', 'main_category', 'deadline', 'goal',
       'launched', 'state', 'backers', 'country', 'usd_pledged', 'currency',
       'pledged'],
      dtype='object')


In [ ]:
print(ks_project.dtypes)

id                 int64
name              object
category          object
main_category     object
deadline          object
goal             float64
launched          object
state             object
backers            int64
country           object
usd_pledged      float64
currency          object
pledged          float64
dtype: object


In [ ]:
print(type(ks_project))

<class 'pandas.core.frame.DataFrame'>


### 3- Do any of the columns contain null values?

In [ ]:
ks_project.isnull().values.any()

True

In [ ]:
ks_project.isnull()

id   name  category  main_category  deadline   goal  launched  \
0      False  False     False          False     False  False     False   
1      False  False     False          False     False  False     False   
2      False  False     False          False     False  False     False   
3      False  False     False          False     False  False     False   
4      False  False     False          False     False  False     False   
...      ...    ...       ...            ...       ...    ...       ...   
65413  False  False     False          False     False  False     False   
65414  False  False     False          False     False  False     False   
65415  False  False     False          False     False  False     False   
65416  False  False     False          False     False  False     False   
65417  False  False     False          False     False  False     False   

       state  backers  country  usd_pledged  currency  pledged  
0      False    False    False        False     False    False  
1      False    False    False        False     False    False  
2      False    False    False        False     False    False  
3      False    False    False        False     False    False  
4      False    False    False        False     False    False  
...      ...      ...      ...          ...       ...      ...  
65413  False    False    False        False     False    False  
65414  False    False    False        False     False    False  
65415  False    False    False        False     False    False  
65416  False    False    False        False     False    False  
65417  False    False    False        False     False    False  

[65418 rows x 13 columns]

### 4- Find all successful documentary projects and sort them by the amount pledged. Print the top 10 highest pledges.

In [ ]:
ks_project['main_category']

0        Film & Video
1        Film & Video
2          Publishing
3             Theater
4        Film & Video
             ...     
65413      Technology
65414      Technology
65415          Design
65416      Technology
65417          Design
Name: main_category, Length: 65418, dtype: object

In [ ]:
ks_project['category']

0           Documentary
1           Documentary
2            Nonfiction
3               Theater
4          Film & Video
              ...      
65413             Sound
65414           Gadgets
65415    Product Design
65416        Technology
65417    Product Design
Name: category, Length: 65418, dtype: object

In [ ]:
success_doc_pro = ks_project[(ks_project["category"] == "Documentary") & (ks_project["state"] == "successful")]
print(success_doc_pro)

               id                                               name  \
1      1100844465                                       daily digest   
9      1125367305            Meeting Andrei Tarkovsky - Film Website   
24     1216330417  Do It Again: One Man's Quest to Reunite the Kinks   
39     1039240602       The Film Dailies - Live From The Walnut Room   
40     1211984590                               'Mister Rogers & Me'   
...           ...                                                ...   
64092  1205743261         AgriCultured: The Lives of Organic Farmers   
64096  1334836761                           The Last Ring Home on TV   
64117  1175698090                                 BlackBook Medellin   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   
64259  1287371551                                   Connections Film   

          category main_category             deadline     goal  \
1      Documentary  Film & Video  2009-06-01 02:20:00    700.0   
9  

In [ ]:
success_doc_pro

id                                               name  \
1      1100844465                                       daily digest   
9      1125367305            Meeting Andrei Tarkovsky - Film Website   
24     1216330417  Do It Again: One Man's Quest to Reunite the Kinks   
39     1039240602       The Film Dailies - Live From The Walnut Room   
40     1211984590                               'Mister Rogers & Me'   
...           ...                                                ...   
64092  1205743261         AgriCultured: The Lives of Organic Farmers   
64096  1334836761                           The Last Ring Home on TV   
64117  1175698090                                 BlackBook Medellin   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   
64259  1287371551                                   Connections Film   

          category main_category             deadline     goal  \
1      Documentary  Film & Video  2009-06-01 02:20:00    700.0   
9      Documentary  Film & Video  2009-06-21 02:31:00   1250.0   
24     Documentary  Film & Video  2009-08-10 21:26:00   5500.0   
39     Documentary  Film & Video  2009-09-15 01:00:00    500.0   
40     Documentary  Film & Video  2009-09-20 05:59:00  10000.0   
...            ...           ...                  ...      ...   
64092  Documentary  Film & Video  2016-11-27 22:36:26   1000.0   
64096  Documentary  Film & Video  2016-11-28 00:59:00   6000.0   
64117  Documentary  Film & Video  2016-11-28 19:27:52   5000.0   
64129  Documentary  Film & Video  2016-11-29 08:00:00   3000.0   
64259  Documentary  Film & Video  2016-12-01 23:28:43  25000.0   

                  launched       state  backers country  usd_pledged currency  \
1      2009-05-04 21:14:28  successful       14      US       700.00      USD   
9      2009-05-30 06:27:42  successful       60      US      2721.00      USD   
24     2009-05-17 05:55:13  successful       79      US      6632.32      USD   
39     2009-08-15 19:52:42  successful        8      US       730.00      USD   
40     2009-08-19 06:16:34  successful      176      US     10795.70      USD   
...                    ...         ...      ...     ...          ...      ...   
64092  2016-10-28 22:36:26  successful       10      US       500.00      USD   
64096  2016-10-20 07:56:50  successful      176      US      4055.00      USD   
64117  2016-10-19 19:27:52  successful       56      US      1106.00      USD   
64129  2016-11-13 20:15:35  successful       74      US         0.00      USD   
64259  2016-10-27 23:28:43  successful      179      US       300.00      USD   

        pledged  
1        700.00  
9       2721.00  
24      6632.32  
39       730.00  
40     10795.70  
...         ...  
64092   1000.00  
64096  40830.00  
64117   5437.00  
64129   5000.00  
64259  25581.00  

[1087 rows x 13 columns]

In [ ]:
success_doc_pro.sort_values(by=['usd_pledged'])

id                                               name  \
64129  1040239768  The Protestors at Standing Rock - Their Testim...   
63047  1077980678                              Lives of The Homeless   
62397  1073535790                                    London Recruits   
31857  1025733989                                           Dominoes   
61581   100358406                                 Charlie vs Goliath   
...           ...                                                ...   
28003  1205913248                                     ILEANA'S SMILE   
12127  1326363492                                   The Culture High   
22237   122290911                          The Killing of Tony Blair   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
11731  1078982332                BRIDEGROOM - An American Love Story   

          category main_category             deadline      goal  \
64129  Documentary  Film & Video  2016-11-29 08:00:00    3000.0   
63047  Documentary  Film & Video  2016-10-25 00:51:09     100.0   
62397  Documentary  Film & Video  2016-10-02 01:00:00    8000.0   
31857  Documentary  Film & Video  2014-08-21 03:25:28      10.0   
61581  Documentary  Film & Video  2016-09-02 02:00:00   40000.0   
...            ...           ...                  ...       ...   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   

                  launched       state  backers country  usd_pledged currency  \
64129  2016-11-13 20:15:35  successful       74      US        0.000      USD   
63047  2016-10-04 00:51:09  successful        1      US        0.000      USD   
62397  2016-09-08 09:02:17  successful      103      GB        0.000      GBP   
31857  2014-08-06 03:25:28  successful        6      CA       18.312      CAD   
61581  2016-08-03 21:54:32  successful      267      US       25.000      USD   
...                    ...         ...      ...     ...          ...      ...   
28003  2014-04-21 09:05:00  successful     1173      US   177421.000      USD   
12127  2012-06-22 01:40:20  successful     3448      US   240022.000      USD   
22237  2013-08-28 18:17:14  successful     4459      GB   255150.000      GBP   
24166  2013-10-30 21:06:34  successful     4245      US   313158.000      USD   
11731  2012-06-20 01:00:28  successful     6508      US   384376.000      USD   

        pledged  
64129    5000.0  
63047     100.0  
62397    8603.0  
31857      20.0  
61581   40546.0  
...         ...  
28003  177421.0  
12127  240022.0  
22237  163891.0  
24166  313158.0  
11731  384376.0  

[1087 rows x 13 columns]

In [ ]:
pledged_amount = success_doc_pro.sort_values('usd_pledged', ascending=False)
pledged_amount

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
22237   122290911                          The Killing of Tony Blair   
12127  1326363492                                   The Culture High   
28003  1205913248                                     ILEANA'S SMILE   
...           ...                                                ...   
61581   100358406                                 Charlie vs Goliath   
31857  1025733989                                           Dominoes   
63047  1077980678                              Lives of The Homeless   
62397  1073535790                                    London Recruits   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
...            ...           ...                  ...       ...   
61581  Documentary  Film & Video  2016-09-02 02:00:00   40000.0   
31857  Documentary  Film & Video  2014-08-21 03:25:28      10.0   
63047  Documentary  Film & Video  2016-10-25 00:51:09     100.0   
62397  Documentary  Film & Video  2016-10-02 01:00:00    8000.0   
64129  Documentary  Film & Video  2016-11-29 08:00:00    3000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US   384376.000      USD   
24166  2013-10-30 21:06:34  successful     4245      US   313158.000      USD   
22237  2013-08-28 18:17:14  successful     4459      GB   255150.000      GBP   
12127  2012-06-22 01:40:20  successful     3448      US   240022.000      USD   
28003  2014-04-21 09:05:00  successful     1173      US   177421.000      USD   
...                    ...         ...      ...     ...          ...      ...   
61581  2016-08-03 21:54:32  successful      267      US       25.000      USD   
31857  2014-08-06 03:25:28  successful        6      CA       18.312      CAD   
63047  2016-10-04 00:51:09  successful        1      US        0.000      USD   
62397  2016-09-08 09:02:17  successful      103      GB        0.000      GBP   
64129  2016-11-13 20:15:35  successful       74      US        0.000      USD   

        pledged  
11731  384376.0  
24166  313158.0  
22237  163891.0  
12127  240022.0  
28003  177421.0  
...         ...  
61581   40546.0  
31857      20.0  
63047     100.0  
62397    8603.0  
64129    5000.0  

[1087 rows x 13 columns]

In [ ]:
pledged_amount.head(10)

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
22237   122290911                          The Killing of Tony Blair   
12127  1326363492                                   The Culture High   
28003  1205913248                                     ILEANA'S SMILE   
27402  1135999291                             Rewind To Fast-Forward   
16243  1061853312                               Linsanity: the Movie   
27268  1228065368                  El Voto Hispano | The Documentary   
43820  1194449879  PlantPure Nation, a Documentary with Dr. T. Co...   
55322  1264072338                                   BREXIT THE MOVIE   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
27402  Documentary  Film & Video  2014-05-01 07:59:00  137000.0   
16243  Documentary  Film & Video  2013-02-16 00:00:00  117000.0   
27268  Documentary  Film & Video  2014-04-27 08:15:23  158000.0   
43820  Documentary  Film & Video  2015-05-14 17:41:41  150000.0   
55322  Documentary  Film & Video  2016-03-02 01:02:47  100000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US     384376.0      USD   
24166  2013-10-30 21:06:34  successful     4245      US     313158.0      USD   
22237  2013-08-28 18:17:14  successful     4459      GB     255150.0      GBP   
12127  2012-06-22 01:40:20  successful     3448      US     240022.0      USD   
28003  2014-04-21 09:05:00  successful     1173      US     177421.0      USD   
27402  2014-03-30 13:35:31  successful     4395      US     176746.0      USD   
16243  2013-01-15 19:46:34  successful     2102      US     167917.0      USD   
27268  2014-03-18 07:15:23  successful      461      US     165401.0      USD   
43820  2015-03-15 16:41:41  successful     1099      US     163930.0      USD   
55322  2016-02-01 01:02:47  successful     1685      GB     162393.0      GBP   

        pledged  
11731  384376.0  
24166  313158.0  
22237  163891.0  
12127  240022.0  
28003  177421.0  
27402  176746.0  
16243  167917.0  
27268  165401.0  
43820  163930.0  
55322  114000.0

##### - The top 10 highest pledges.

### 5- Create a new column named average_per_backer and set its value to the total amount pledged / number of backers.

In [ ]:
pledged_amount["average_per_backer"] = pledged_amount["usd_pledged"] / pledged_amount["backers"]
pledged_amount

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
22237   122290911                          The Killing of Tony Blair   
12127  1326363492                                   The Culture High   
28003  1205913248                                     ILEANA'S SMILE   
...           ...                                                ...   
61581   100358406                                 Charlie vs Goliath   
31857  1025733989                                           Dominoes   
63047  1077980678                              Lives of The Homeless   
62397  1073535790                                    London Recruits   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
...            ...           ...                  ...       ...   
61581  Documentary  Film & Video  2016-09-02 02:00:00   40000.0   
31857  Documentary  Film & Video  2014-08-21 03:25:28      10.0   
63047  Documentary  Film & Video  2016-10-25 00:51:09     100.0   
62397  Documentary  Film & Video  2016-10-02 01:00:00    8000.0   
64129  Documentary  Film & Video  2016-11-29 08:00:00    3000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US   384376.000      USD   
24166  2013-10-30 21:06:34  successful     4245      US   313158.000      USD   
22237  2013-08-28 18:17:14  successful     4459      GB   255150.000      GBP   
12127  2012-06-22 01:40:20  successful     3448      US   240022.000      USD   
28003  2014-04-21 09:05:00  successful     1173      US   177421.000      USD   
...                    ...         ...      ...     ...          ...      ...   
61581  2016-08-03 21:54:32  successful      267      US       25.000      USD   
31857  2014-08-06 03:25:28  successful        6      CA       18.312      CAD   
63047  2016-10-04 00:51:09  successful        1      US        0.000      USD   
62397  2016-09-08 09:02:17  successful      103      GB        0.000      GBP   
64129  2016-11-13 20:15:35  successful       74      US        0.000      USD   

        pledged  average_per_backer  
11731  384376.0           59.062077  
24166  313158.0           73.771025  
22237  163891.0           57.221350  
12127  240022.0           69.611949  
28003  177421.0          151.254049  
...         ...                 ...  
61581   40546.0            0.093633  
31857      20.0            3.052000  
63047     100.0            0.000000  
62397    8603.0            0.000000  
64129    5000.0            0.000000  

[1087 rows x 14 columns]

In [ ]:
pledged_amount["usd_pledged"]

11731    384376.000
24166    313158.000
22237    255150.000
12127    240022.000
28003    177421.000
            ...    
61581        25.000
31857        18.312
63047         0.000
62397         0.000
64129         0.000
Name: usd_pledged, Length: 1087, dtype: float64

### 6- What happened to the rows with 0 backers? How can this be dealt with?

In [ ]:
pledged_amount.shape

(1087, 14)

### 7- Drop all rows with 0 backers then repeat the previous exercise.

In [ ]:
pledged_amount.dropna()

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
22237   122290911                          The Killing of Tony Blair   
12127  1326363492                                   The Culture High   
28003  1205913248                                     ILEANA'S SMILE   
...           ...                                                ...   
61581   100358406                                 Charlie vs Goliath   
31857  1025733989                                           Dominoes   
63047  1077980678                              Lives of The Homeless   
62397  1073535790                                    London Recruits   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
...            ...           ...                  ...       ...   
61581  Documentary  Film & Video  2016-09-02 02:00:00   40000.0   
31857  Documentary  Film & Video  2014-08-21 03:25:28      10.0   
63047  Documentary  Film & Video  2016-10-25 00:51:09     100.0   
62397  Documentary  Film & Video  2016-10-02 01:00:00    8000.0   
64129  Documentary  Film & Video  2016-11-29 08:00:00    3000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US   384376.000      USD   
24166  2013-10-30 21:06:34  successful     4245      US   313158.000      USD   
22237  2013-08-28 18:17:14  successful     4459      GB   255150.000      GBP   
12127  2012-06-22 01:40:20  successful     3448      US   240022.000      USD   
28003  2014-04-21 09:05:00  successful     1173      US   177421.000      USD   
...                    ...         ...      ...     ...          ...      ...   
61581  2016-08-03 21:54:32  successful      267      US       25.000      USD   
31857  2014-08-06 03:25:28  successful        6      CA       18.312      CAD   
63047  2016-10-04 00:51:09  successful        1      US        0.000      USD   
62397  2016-09-08 09:02:17  successful      103      GB        0.000      GBP   
64129  2016-11-13 20:15:35  successful       74      US        0.000      USD   

        pledged  average_per_backer  
11731  384376.0           59.062077  
24166  313158.0           73.771025  
22237  163891.0           57.221350  
12127  240022.0           69.611949  
28003  177421.0          151.254049  
...         ...                 ...  
61581   40546.0            0.093633  
31857      20.0            3.052000  
63047     100.0            0.000000  
62397    8603.0            0.000000  
64129    5000.0            0.000000  

[1087 rows x 14 columns]

In [ ]:
pledged_amount.shape

(1087, 14)

### 8- Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.

In [ ]:
index = list('abcdefghij')
primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
fibonacci = [1, 1, 2, 3, 5, 8, 13, 21, 34, 55]

df = pd.DataFrame({"Primes": primes, "Fibonacci": fibonacci}, index = index)
print(df)

   Primes  Fibonacci
a       2          1
b       3          1
c       5          2
d       7          3
e      11          5
f      13          8
g      17         13
h      19         21
i      23         34
j      29         55


### 9- Merge these two DataFrames into a single DataFrame: nycflights13.

Please use the following dataset for merging purposes:

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/flights.csv

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/airlines.csv

In [ ]:
flights = pd.read_csv(r"C:\Users\AAygun\Documents\GitHub\Data Sets\Magnimind\flights.csv")
airlines = pd.read_csv(r"C:\Users\AAygun\Documents\GitHub\Data Sets\Magnimind\airlines.csv")

In [ ]:
flights

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute            time_hour  
0          227.0      1400     5      15  2013-01-01 05:00:00  
1          227.0      1416     5      29  2013-01-01 05:00:00  
2          160.0      1089     5      40  2013-01-01 05:00:00  
3          183.0      1576     5      45  2013-01-01 05:00:00  
4          116.0       762     6       0  2013-01-01 06:00:00  
...          ...       ...   ...     ...                  ...  
336771       NaN       213    14      55  2013-09-30 14:00:00  
336772       NaN       198    22       0  2013-09-30 22:00:00  
336773       NaN       764    12      10  2013-09-30 12:00:00  
336774       NaN       419    11      59  2013-09-30 11:00:00  
336775       NaN       431     8      40  2013-09-30 08:00:00  

[336776 rows x 19 columns]

In [ ]:
airlines

carrier                         name
0       9E            Endeavor Air Inc.
1       AA       American Airlines Inc.
2       AS         Alaska Airlines Inc.
3       B6              JetBlue Airways
4       DL         Delta Air Lines Inc.
5       EV     ExpressJet Airlines Inc.
6       F9       Frontier Airlines Inc.
7       FL  AirTran Airways Corporation
8       HA       Hawaiian Airlines Inc.
9       MQ                    Envoy Air
10      OO        SkyWest Airlines Inc.
11      UA        United Air Lines Inc.
12      US              US Airways Inc.
13      VX               Virgin America
14      WN       Southwest Airlines Co.
15      YV           Mesa Airlines Inc.

In [ ]:
nycflights13 = pd.merge(flights, airlines)
nycflights13

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     554.0             558       -4.0     740.0   
3       2013      1    1     558.0             600       -2.0     924.0   
4       2013      1    1     558.0             600       -2.0     923.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   19    1759.0            1805       -6.0    1946.0   
336772  2013      9   20    1758.0            1805       -7.0    1929.0   
336773  2013      9   22    1759.0            1805       -6.0    1945.0   
336774  2013      9   23    1759.0            1805       -6.0    1935.0   
336775  2013      9   24    1751.0            1805      -14.0    1937.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  728       12.0      UA    1696  N39463    EWR  ORD   
3                  917        7.0      UA     194  N29129    JFK  LAX   
4                  937      -14.0      UA    1124  N53441    EWR  SFO   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1953       -7.0      OO    5568  N760SK    LGA  CLE   
336772            1953      -24.0      OO    5568  N766SK    LGA  CLE   
336773            1953       -8.0      OO    5568  N772SK    LGA  CLE   
336774            1953      -18.0      OO    5568  N776SK    LGA  CLE   
336775            1953      -16.0      OO    5568  N785SK    LGA  CLE   

        air_time  distance  hour  minute            time_hour  \
0          227.0      1400     5      15  2013-01-01 05:00:00   
1          227.0      1416     5      29  2013-01-01 05:00:00   
2          150.0       719     5      58  2013-01-01 05:00:00   
3          345.0      2475     6       0  2013-01-01 06:00:00   
4          361.0      2565     6       0  2013-01-01 06:00:00   
...          ...       ...   ...     ...                  ...   
336771      64.0       419    18       5  2013-09-19 18:00:00   
336772      60.0       419    18       5  2013-09-20 18:00:00   
336773      67.0       419    18       5  2013-09-22 18:00:00   
336774      68.0       419    18       5  2013-09-23 18:00:00   
336775      67.0       419    18       5  2013-09-24 18:00:00   

                         name  
0       United Air Lines Inc.  
1       United Air Lines Inc.  
2       United Air Lines Inc.  
3       United Air Lines Inc.  
4       United Air Lines Inc.  
...                       ...  
336771  SkyWest Airlines Inc.  
336772  SkyWest Airlines Inc.  
336773  SkyWest Airlines Inc.  
336774  SkyWest Airlines Inc.  
336775  SkyWest Airlines Inc.  

[336776 rows x 20 columns]

In [ ]:
nycflights13.isnull().any()

In [ ]:
nycflights13.notnull()

In [ ]:
new_dataframe = nycflights13.dropna(axis = 0, how ='any')

In [ ]:
nycflights13.isnull().any()

### 11- Merge these two DataFrames into a single DataFrame called census.

Please use the following dataset for merging purposes:

'https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv'

'https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv'

In [ ]:
state_populations = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv')
census_divisions = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv')

In [ ]:
state_populations

In [ ]:
census_divisions

In [ ]:
census = pd.merge(state_populations, census_divisions)
census

### 12- Re-shape census so that one column contains all population measures, and another the year attributes.

In [ ]:
census.shape

In [ ]:
df = census.reset_index()
year_1 = pd.melt(census, id_vars=["state", "region", "division"],  value_vars=["2010", "2011", "2012", "2013", "2014", "2015", "2016"])
year_1

In [ ]:
year_2 = year_1.rename(columns={"variable": "year", "value": "population measures"})

In [ ]:
year_2

In [ ]:
grouped_by_year = year_2.groupby('year')
grouped_by_year.first()

### 15- Create a crosstab between state and category showing amount pledged as the value.

Please refer to the following link for this practice:

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv

In [ ]:
ksproject=pd.read_csv(r"C:\Users\AAygun\Documents\GitHub\Data Sets\Magnimind\ksprojects.csv")

In [ ]:
ksproject.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

In [ ]:
pd.crosstab(ksproject.state, ksproject.category)

category    3D Printing  Academic  Accessories  Action  Animals  Animation  \
state                                                                        
canceled             17        21           48      14        2         41   
failed               59        86          238      79       31        245   
live                  3         3           16       1        0          8   
successful           43        43          121      16       11        134   
suspended             2         4            2       0        1          1   
undefined             1         0            1       0        0          0   

category    Anthologies  Apparel  Apps  Architecture  ...  Wearables  Weaving  \
state                                                 ...                       
canceled              4      127   166            10  ...         25        3   
failed               34      684   784            86  ...         71        8   
live                  6       37    28             4  ...          7        0   
successful           83      211    55            26  ...         57        3   
suspended             0        5     5             2  ...          2        1   
undefined             0        0     0             0  ...          0        0   

category    Web  Webcomics  Webseries  Woodworking  Workshops  World Music  \
state                                                                        
canceled    114          4        119           10          0           31   
failed      634         28        656          125         12          174   
live         23          3          5            4          1            2   
successful   73         55        313           38          8          168   
suspended     6          0          6            2          1            1   
undefined     0          0          0            0          0            0   

category    Young Adult  Zines  
state                           
canceled             16      6  
failed               78     26  
live                  1      1  
successful           23     17  
suspended             0      0  
undefined             0      0  

[6 rows x 158 columns]

# OFFICE HOUR SOLUTIONS

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv(r"https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv")
df.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

In [4]:
print(len(df))

len(df) == df.shape[0]

65418


True

In [5]:
df.dtypes

id                 int64
name              object
category          object
main_category     object
deadline          object
goal             float64
launched          object
state             object
backers            int64
country           object
usd_pledged      float64
currency          object
pledged          float64
dtype: object

In [6]:
df.isnull().sum()

id                 0
name               0
category           0
main_category      0
deadline           0
goal               0
launched           0
state              0
backers            0
country            0
usd_pledged      745
currency           0
pledged            0
dtype: int64

In [7]:
df.isnull().sum()/len(df)*100 # percentage of null values

id               0.00000
name             0.00000
category         0.00000
main_category    0.00000
deadline         0.00000
goal             0.00000
launched         0.00000
state            0.00000
backers          0.00000
country          0.00000
usd_pledged      1.13883
currency         0.00000
pledged          0.00000
dtype: float64

In [8]:
# Find all successful documentaty projects and sort them by the amount of pledged.
# Print the top 10

df.loc[df['state'] == 'successful'].sort_values(by='pledged').head(10)

df[df['state'] == 'successful'].sort_values(by='usd_pledged', ascending=False).head()



id                                            name  \
12253  1033978702          OUYA: A New Kind of Video Game Console   
26917  1152735261  Pono Music - Where Your Soul Rediscovers Music   
45290  1074800880               Bloodstained: Ritual of the Night   
46419  1045500705                         Zombicide: Black Plague   
59897   132311747                                Massive Darkness   

              category main_category             deadline      goal  \
12253  Gaming Hardware         Games  2012-08-09 07:00:00  950000.0   
26917            Sound    Technology  2014-04-15 19:30:42  800000.0   
45290      Video Games         Games  2015-06-13 05:00:00  500000.0   
46419   Tabletop Games         Games  2015-07-07 02:00:00  125000.0   
59897   Tabletop Games         Games  2016-07-08 01:00:00  200000.0   

                  launched       state  backers country  usd_pledged currency  \
12253  2012-07-10 14:44:41  successful    63416      US    8596480.0      USD   
26917  2014-03-11 18:30:42  successful    18220      US    6225360.0      USD   
45290  2015-05-11 18:58:41  successful    64867      US    5545990.0      USD   
46419  2015-06-08 20:59:38  successful    20915      US    4079200.0      USD   
59897  2016-06-07 21:02:31  successful    22361      US    3560640.0      USD   

         pledged  
12253  8596480.0  
26917  6225360.0  
45290  5545990.0  
46419  4079200.0  
59897  3560640.0

In [9]:
# Create a new column named average_per_baker
# and set its value to the total amount pledged / number of backers

df['average_per_backer'] = df['pledged']/df['backers']
df.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  average_per_backer  
0    40.00           10.000000  
1   700.00           50.000000  
2  1367.00           42.718750  
3  6575.00          273.958333  
4   501.66           22.802727

In [10]:
df['average_per_backer'].isnull().sum()

9282

In [11]:
df['pledged']/df.loc[df['backers'] == 0, 'backers']

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
         ..
65413   NaN
65414   NaN
65415   NaN
65416   NaN
65417   NaN
Length: 65418, dtype: float64

In [14]:
df['average_per_backer'] = df['pledged'].div(df['backers'].values)
df.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  average_per_backer  
0    40.00           10.000000  
1   700.00           50.000000  
2  1367.00           42.718750  
3  6575.00          273.958333  
4   501.66           22.802727

In [15]:
#What happened to the rows with 0 backers?
#How can this be dealt with

df['average_per_backer'] = df.loc[df['backers']!=0, 'pledged'] / df.loc[df['backers']!=0, 'backers']
df.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  average_per_backer  
0    40.00           10.000000  
1   700.00           50.000000  
2  1367.00           42.718750  
3  6575.00          273.958333  
4   501.66           22.802727

In [8]:
import numpy as np
import pandas as pd

In [20]:
df['average_per_backer'] = np.where(df['backers']!=0, df['pledged'] / df['backers'], np.nan)
df.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  average_per_backer  
0    40.00           10.000000  
1   700.00           50.000000  
2  1367.00           42.718750  
3  6575.00          273.958333  
4   501.66           22.802727

In [22]:
# np.where(condition, value1, value2)

### Solution 2

### 8- Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.

In [5]:
Prime_number = []

for i in range(100):
  L = []
  for j in range(2,i):
    if i%j == 0:
      L.append(i)
  if len(L) == 0 and i!=0 and i!=1:
    Prime_number.append(i)

Prime_number[:10]

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]

In [6]:
fibonacci_numbers = [0,1]

i = 0
while i<8:
  fibonacci_numbers.append(fibonacci_numbers[-1]+fibonacci_numbers[-2])
  i += 1

fibonacci_numbers  

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [10]:
import string
string.ascii_uppercase

'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [13]:
list(string.ascii_uppercase[:10])

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J']

In [14]:

df = pd.DataFrame({'Prime_number' : Prime_number[:10],
                   'fibonacci_numbers' : fibonacci_numbers},
                  index=list(string.ascii_uppercase[:10]))
df

Prime_number  fibonacci_numbers
A             2                  0
B             3                  1
C             5                  1
D             7                  2
E            11                  3
F            13                  5
G            17                  8
H            19                 13
I            23                 21
J            29                 34

In [15]:
df.shape

(10, 2)

## SOLUTION 3

Merge these two DataFrames into a single DataFrame: nycflights13.
Please use the following dataset for merging purposes:

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/flights.csv

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/airlines.csv

In [3]:
import pandas as pd

In [4]:
flights = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/flights.csv')
airlines = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/airlines.csv')

In [5]:
display(flights.head())
display(airlines.head())

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour  
0      1400     5      15  2013-01-01 05:00:00  
1      1416     5      29  2013-01-01 05:00:00  
2      1089     5      40  2013-01-01 05:00:00  
3      1576     5      45  2013-01-01 05:00:00  
4       762     6       0  2013-01-01 06:00:00

carrier                    name
0      9E       Endeavor Air Inc.
1      AA  American Airlines Inc.
2      AS    Alaska Airlines Inc.
3      B6         JetBlue Airways
4      DL    Delta Air Lines Inc.

In [6]:
flights['carrier'].value_counts()

UA    58665
B6    54635
EV    54173
DL    48110
AA    32729
MQ    26397
US    20536
9E    18460
WN    12275
VX     5162
FL     3260
AS      714
F9      685
YV      601
HA      342
OO       32
Name: carrier, dtype: int64

In [7]:
airlines['carrier'].value_counts()

9E    1
AA    1
AS    1
B6    1
DL    1
EV    1
F9    1
FL    1
HA    1
MQ    1
OO    1
UA    1
US    1
VX    1
WN    1
YV    1
Name: carrier, dtype: int64

In [10]:
pd. merge(left=flights, right=airlines, on='carrier', how='outer')[['carrier', 'name']].value_counts()

carrier  name                       
UA       United Air Lines Inc.          58665
B6       JetBlue Airways                54635
EV       ExpressJet Airlines Inc.       54173
DL       Delta Air Lines Inc.           48110
AA       American Airlines Inc.         32729
MQ       Envoy Air                      26397
US       US Airways Inc.                20536
9E       Endeavor Air Inc.              18460
WN       Southwest Airlines Co.         12275
VX       Virgin America                  5162
FL       AirTran Airways Corporation     3260
AS       Alaska Airlines Inc.             714
F9       Frontier Airlines Inc.           685
YV       Mesa Airlines Inc.               601
HA       Hawaiian Airlines Inc.           342
OO       SkyWest Airlines Inc.             32
dtype: int64

In [11]:
nycflights13 = pd.merge(flights, airlines, on='carrier', how='inner')
nycflights13 = pd.merge(flights, airlines, on='carrier', how='outer')

display(nycflights13.sample(n=5, axis=0))

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
272968  2013      9    7    1134.0            1135       -1.0    1236.0   
320096  2013     11    4    1854.0            1900       -6.0    2144.0   
236482  2013      7   15    1634.0            1643       -9.0    1852.0   
136928  2013      8    1    2112.0            1936       96.0       9.0   
138921  2013      8   14     927.0             930       -3.0    1034.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
272968            1300      -24.0      MQ    3461  N505MQ    LGA  BNA   
320096            2124       20.0      9E    3445  N936XJ    LGA  DSM   
236482            1907      -15.0      EV    5601  N720EV    LGA  CLT   
136928            2233       96.0      B6    1271  N613JB    LGA  FLL   
138921            1048      -14.0      B6     586  N324JB    JFK  ROC   

        air_time  distance  hour  minute            time_hour  \
272968      96.0       764    11      35  2013-09-07 11:00:00   
320096     164.0      1031    19       0  2013-11-04 19:00:00   
236482      88.0       544    16      43  2013-07-15 16:00:00   
136928     157.0      1076    19      36  2013-08-01 19:00:00   
138921      51.0       264     9      30  2013-08-14 09:00:00   

                            name  
272968                 Envoy Air  
320096         Endeavor Air Inc.  
236482  ExpressJet Airlines Inc.  
136928           JetBlue Airways  
138921           JetBlue Airways

In [12]:
display(nycflights13.sample(n=10,axis=1))

dep_time  day  flight  distance  arr_time  sched_arr_time carrier  \
0          517.0    1    1545      1400     830.0             819      UA   
1          533.0    1    1714      1416     850.0             830      UA   
2          554.0    1    1696       719     740.0             728      UA   
3          558.0    1     194      2475     924.0             917      UA   
4          558.0    1    1124      2565     923.0             937      UA   
...          ...  ...     ...       ...       ...             ...     ...   
336771    1759.0   19    5568       419    1946.0            1953      OO   
336772    1758.0   20    5568       419    1929.0            1953      OO   
336773    1759.0   22    5568       419    1945.0            1953      OO   
336774    1759.0   23    5568       419    1935.0            1953      OO   
336775    1751.0   24    5568       419    1937.0            1953      OO   

        month            time_hour  sched_dep_time  
0           1  2013-01-01 05:00:00             515  
1           1  2013-01-01 05:00:00             529  
2           1  2013-01-01 05:00:00             558  
3           1  2013-01-01 06:00:00             600  
4           1  2013-01-01 06:00:00             600  
...       ...                  ...             ...  
336771      9  2013-09-19 18:00:00            1805  
336772      9  2013-09-20 18:00:00            1805  
336773      9  2013-09-22 18:00:00            1805  
336774      9  2013-09-23 18:00:00            1805  
336775      9  2013-09-24 18:00:00            1805  

[336776 rows x 10 columns]

In [14]:
#join() is used to combine two Dataframes on the index
# but not on columns whereas merge is primarily used to
# specify the columns you wanted to join on

#In other word Merge combines Dataframes based on values in shared columns.

nycflights13 = flights.set_index('carrier').join(airlines.set_index('carrier'))
nycflights13

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
carrier                                                                    
9E       2013      1    1     810.0             810        0.0    1048.0   
9E       2013      1    1    1451.0            1500       -9.0    1634.0   
9E       2013      1    1    1452.0            1455       -3.0    1637.0   
9E       2013      1    1    1454.0            1500       -6.0    1635.0   
9E       2013      1    1    1507.0            1515       -8.0    1651.0   
...       ...    ...  ...       ...             ...        ...       ...   
YV       2013      9   29    1650.0            1629       21.0    1810.0   
YV       2013      9   29    1801.0            1735       26.0    1957.0   
YV       2013      9   30    1643.0            1629       14.0    1800.0   
YV       2013      9   30    1731.0            1735       -4.0    1921.0   
YV       2013      9   30    2000.0            2000        0.0    2207.0   

         sched_arr_time  arr_delay  flight tailnum origin dest  air_time  \
carrier                                                                    
9E                 1037       11.0    3538  N915XJ    JFK  MSP     189.0   
9E                 1636       -2.0    4105  N8444F    JFK  IAD      57.0   
9E                 1639       -2.0    3295  N920XJ    JFK  BUF      68.0   
9E                 1636       -1.0    3843  N8409N    JFK  SYR      57.0   
9E                 1656       -5.0    3792  N8631E    JFK  ROC      66.0   
...                 ...        ...     ...     ...    ...  ...       ...   
YV                 1750       20.0    3771  N518LR    LGA  IAD      41.0   
YV                 1946       11.0    2751  N932LR    LGA  CLT      79.0   
YV                 1750       10.0    3771  N510MJ    LGA  IAD      42.0   
YV                 1946      -25.0    2751  N905FJ    LGA  CLT      78.0   
YV                 2211       -4.0    2677  N924FJ    LGA  CLT      75.0   

         distance  hour  minute            time_hour                name  
carrier                                                                   
9E           1029     8      10  2013-01-01 08:00:00   Endeavor Air Inc.  
9E            228    15       0  2013-01-01 15:00:00   Endeavor Air Inc.  
9E            301    14      55  2013-01-01 14:00:00   Endeavor Air Inc.  
9E            209    15       0  2013-01-01 15:00:00   Endeavor Air Inc.  
9E            264    15      15  2013-01-01 15:00:00   Endeavor Air Inc.  
...           ...   ...     ...                  ...                 ...  
YV            229    16      29  2013-09-29 16:00:00  Mesa Airlines Inc.  
YV            544    17      35  2013-09-29 17:00:00  Mesa Airlines Inc.  
YV            229    16      29  2013-09-30 16:00:00  Mesa Airlines Inc.  
YV            544    17      35  2013-09-30 17:00:00  Mesa Airlines Inc.  
YV            544    20       0  2013-09-30 20:00:00  Mesa Airlines Inc.  

[336776 rows x 19 columns]

In [17]:
pd.concat([flights.set_index('carrier'), airlines.set_index('carrier')])

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
carrier                                                                      
UA       2013.0    1.0  1.0     517.0           515.0        2.0     830.0   
UA       2013.0    1.0  1.0     533.0           529.0        4.0     850.0   
AA       2013.0    1.0  1.0     542.0           540.0        2.0     923.0   
B6       2013.0    1.0  1.0     544.0           545.0       -1.0    1004.0   
DL       2013.0    1.0  1.0     554.0           600.0       -6.0     812.0   
...         ...    ...  ...       ...             ...        ...       ...   
UA          NaN    NaN  NaN       NaN             NaN        NaN       NaN   
US          NaN    NaN  NaN       NaN             NaN        NaN       NaN   
VX          NaN    NaN  NaN       NaN             NaN        NaN       NaN   
WN          NaN    NaN  NaN       NaN             NaN        NaN       NaN   
YV          NaN    NaN  NaN       NaN             NaN        NaN       NaN   

         sched_arr_time  arr_delay  flight tailnum origin dest  air_time  \
carrier                                                                    
UA                819.0       11.0  1545.0  N14228    EWR  IAH     227.0   
UA                830.0       20.0  1714.0  N24211    LGA  IAH     227.0   
AA                850.0       33.0  1141.0  N619AA    JFK  MIA     160.0   
B6               1022.0      -18.0   725.0  N804JB    JFK  BQN     183.0   
DL                837.0      -25.0   461.0  N668DN    LGA  ATL     116.0   
...                 ...        ...     ...     ...    ...  ...       ...   
UA                  NaN        NaN     NaN     NaN    NaN  NaN       NaN   
US                  NaN        NaN     NaN     NaN    NaN  NaN       NaN   
VX                  NaN        NaN     NaN     NaN    NaN  NaN       NaN   
WN                  NaN        NaN     NaN     NaN    NaN  NaN       NaN   
YV                  NaN        NaN     NaN     NaN    NaN  NaN       NaN   

         distance  hour  minute            time_hour                    name  
carrier                                                                       
UA         1400.0   5.0    15.0  2013-01-01 05:00:00                     NaN  
UA         1416.0   5.0    29.0  2013-01-01 05:00:00                     NaN  
AA         1089.0   5.0    40.0  2013-01-01 05:00:00                     NaN  
B6         1576.0   5.0    45.0  2013-01-01 05:00:00                     NaN  
DL          762.0   6.0     0.0  2013-01-01 06:00:00                     NaN  
...           ...   ...     ...                  ...                     ...  
UA            NaN   NaN     NaN                  NaN   United Air Lines Inc.  
US            NaN   NaN     NaN                  NaN         US Airways Inc.  
VX            NaN   NaN     NaN                  NaN          Virgin America  
WN            NaN   NaN     NaN                  NaN  Southwest Airlines Co.  
YV            NaN   NaN     NaN                  NaN      Mesa Airlines Inc.  

[336792 rows x 19 columns]

### Solution 4
Merge these two DataFrames into a single DataFrame called census.
Please use the following dataset for merging purposes:

'https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv'

'https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv'

In [19]:
import pandas as pd
df1 = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv')
df2 = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv')

In [20]:
display(df1.head())
display(df2.head())

state      2010      2011      2012      2013      2014      2015  \
0     Alabama   4785492   4799918   4815960   4829479   4843214   4853875   
1      Alaska    714031    722713    731089    736879    736705    737709   
2     Arizona   6408312   6467163   6549634   6624617   6719993   6817565   
3    Arkansas   2921995   2939493   2950685   2958663   2966912   2977853   
4  California  37332685  37676861  38011074  38335203  38680810  38993940   

       2016  
0   4863300  
1    741894  
2   6931071  
3   2988248  
4  39250017

state     region     division
0    Connecticut  Northeast  New England
1          Maine  Northeast  New England
2  Massachusetts  Northeast  New England
3  New Hampshire  Northeast  New England
4   Rhode Island  Northeast  New England